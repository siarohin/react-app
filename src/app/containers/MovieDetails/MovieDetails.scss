.app-movie-details {
  display: grid;
  grid-template-columns: 2fr 3fr;
  height: calc(100vh - 8 * var(--app-spacing-m));
  column-gap: var(--app-spacing-xs);
  font-size: 1.8rem;
  color: var(--app-theme-grey-light-25);

  &::after {
    content: "";
    position: absolute;
    top: 0;
    left: 0;
    background-color: var(--app-theme-grey-light-900);
    opacity: 0.8;
    height: 100%;
    width: 100%;
    z-index: -1;
  }
}

.app-movie-details__404 {
  grid-column: 1 / span 2;
  align-self: center;
  justify-self: center;
  font-size: 2.8rem;
}

.app-movie-details__poster {
  background-repeat: no-repeat;
  background-size: contain;
}

.app-movie-detils__content {
  display: inline-grid;
  grid-template-rows: repeat(3, minmax(0, max-content));
  grid-template-columns: repeat(2, minmax(0, max-content));
  row-gap: var(--app-spacing-m);
  align-items: center;
}

.app-movie-details__title {
  grid-row: 1 / span 1;
  grid-column: 1 / span 1;
  font-size: 3.2rem;
  letter-spacing: var(--app-letter-spacing);
}

.app-movie-details__rate {
  grid-row: 1 / span 1;
  grid-column: 2 / span 1;
  color: var(--app-theme-data-viz-green);
  font-size: 2.8rem;
  height: 6rem;
  width: 6rem;
  border: 0.1rem solid var(--app-theme-grey-light-25);
  border-radius: 50%;
  display: flex;
  align-items: center;
  justify-content: center;
}

.app-movie-details__pre-description {
  grid-row: 2 / span 1;
  grid-column: 1 / span 2;
  color: var(--app-theme-data-viz-red);
  font-size: 2.8rem;

  &-item {
    margin-right: var(--app-spacing-xl);
  }
}

.app-movie-details__description {
  grid-row: 3 / span 1;
  grid-column: 1 / span 2;
}

.app-movie-details__icon-button {
  *[class*="MuiSvgIcon-root"] {
    font-size: 3.2rem;
    color: var(--app-theme-data-viz-red);
    transform: rotate(90deg);
  }
}
