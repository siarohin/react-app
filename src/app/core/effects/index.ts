import { combineEpics } from "redux-observable";

import { movieEffects } from "./movies.effect";
import { userPreferencesEffects } from "./user-preferences.effect";
import { toastEffects } from "./toast.effect";

export * from "./common.util";

export const effects = combineEpics(movieEffects, userPreferencesEffects, toastEffects);
