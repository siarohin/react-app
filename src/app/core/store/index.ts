export * from "./reducers";
export { SharedModels } from "./shared";
export { MoviesActions, MoviesModels } from "./movies";
export { UserPreferencesActions, UserPreferencesModels } from "./user-preferences";
export { ToastActions } from "./toast";
