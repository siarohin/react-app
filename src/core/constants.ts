/**
 * App name
 */
export const APP_NAME: string = "React movie";

/**
 * Dialog actions
 */
export enum DialogAction {
  CREATE = "Create",
  UPDATE = "Update",
  DELETE = "Delete"
}

/**
 * Dialog titles
 */
export enum DialogTitle {
  CREATE = "Add movie",
  UPDATE = "Edit movie",
  DELETE = "Delete movie"
}

/**
 * Selectable filter options
 */
export enum FilterOptions {
  rating = "Rating",
  releaseDate = "Release date"
}

/**
 * Queries that are used in the filter request
 */
export enum FilterOptionsQuery {
  rating = "vote_average",
  releaseDate = "release_date"
}

/**
 * Filter genres
 */
export const FILTER_GENRES: Array<string> = ["Comedy", "Crime", "Documentary", "Drama", "Horror", "Romance"];

/**
 * Default filter value
 */
export const DEFAULT_GENRE: string = "All";

/**
 * All the backend API endpoint URIs are listed here.
 */
export enum Endpoints {
  // REST APIs:
  AppHost = "http://localhost:4000"
}

/**
 * Movies per page
 */
export const MOVIES_PER_PAGE: number = 30;

/**
 * Sorting order
 */
export const SORTING_ORDER: "asc" | "desc" = "desc";

/**
 * Search option
 */
export const SEARCH_BY: "title" | "genres" = "title";

/**
 * Upsert movie messages enum
 */
export enum UpsertMovieMsg {
  Success = "The movie has been successfully saved.",
  Fail = "The movie could not be saved."
}

/**
 * Delete movie messages enum
 */
export enum DeleteMovieMsg {
  Fail = "The movie could not be deleted.",
  Success = "The movie has been successfully deleted."
}
