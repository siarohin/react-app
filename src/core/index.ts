export * from "./constants";
export * from "./services";
export * from "./effects";
export * from "./store";
export * from "./store-configuration";
