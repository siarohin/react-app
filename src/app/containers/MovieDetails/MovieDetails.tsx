import React, { ComponentType, ReactElement, useEffect, useRef } from "react";
import { useHistory, useParams } from "react-router-dom";
import { Helmet } from "react-helmet";
import { ActionCreatorWithoutPayload } from "@reduxjs/toolkit";
import { connect } from "react-redux";
import isEmpty from "lodash/isEmpty";

import "./MovieDetails.scss";
import { IconButton, SearchIcon, CircularProgress, IMovieDetails } from "../../shared";
import {
  MoviesModels,
  UserPreferencesActions,
  SharedModels,
  State,
  MoviesActions,
  RouterPath,
  MetaTitle,
  Endpoints
} from "../../core";
import { useComponentWillMount } from "../../hooks";
import { getFullYear, getPath } from "../../utils";

const MovieDetails = ({
  movie,
  searchValue,
  dispatch
}: {
  movie: MoviesModels.IMovie;
  searchValue: string;
  dispatch: (action: SharedModels.ActionWithPayload<any>) => ActionCreatorWithoutPayload;
}): ReactElement<IMovieDetails> => {
  const history = useHistory();
  const { id } = useParams<{ id: string }>();

  const containerRef = useRef((null as unknown) as HTMLDivElement);

  useComponentWillMount(() => {
    if (isEmpty(movie)) {
      dispatch(MoviesActions.loadMovieById({ id }));
    }
  });

  if (movie?.hasError) {
    history.push(RouterPath.NotFound);
  }

  useEffect(() => containerRef.current?.scrollIntoView({ behavior: "smooth" }), [movie]);

  const onClose = (): void => {
    const selectedMovie: MoviesModels.IMovie = {} as MoviesModels.IMovie;
    const value: string = searchValue?.trim();

    dispatch(UserPreferencesActions.updateSelectedMovieSuccess({ selectedMovie }));
    history.push(getPath(value));
  };

  return (
    <>
      <Helmet>
        <title>{`${movie?.title} - ${MetaTitle.MovieDetails} in ${getFullYear(movie?.releaseDate)}`}</title>
        <link rel="canonical" href={`${Endpoints.ClientHost}${RouterPath.Movie}/${movie?.id}`} />
      </Helmet>

      <div ref={containerRef} className="app-header__top-right">
        <IconButton className="app-movie-details__icon-button" onClick={onClose}>
          <SearchIcon />
        </IconButton>
      </div>
      <div className="app-movie-details">
        {isEmpty(movie) ? (
          <CircularProgress className="app__progress-bar" size="8rem" color="secondary" />
        ) : (
          <>
            <div
              className="app-movie-details__poster"
              style={{ backgroundImage: "url(" + movie?.posterPath + ")" }}
            ></div>
            <div className="app-movie-detils__content">
              <h3 className="app-movie-details__title">{movie?.title}</h3>
              {movie?.voteAverage ? <div className="app-movie-details__rate">{movie?.voteAverage}</div> : null}
              <div className="app-movie-details__pre-description">
                <span className="app-movie-details__pre-description-item">{getFullYear(movie?.releaseDate)}</span>
                <span className="app-movie-details__pre-description-item">{movie?.runtime} min</span>
              </div>
              <div className="app-movie-details__description">{movie?.overview}</div>
            </div>
          </>
        )}
      </div>
    </>
  );
};

const mapStateToProps = (state: State): { movie: MoviesModels.IMovie; searchValue: string } => {
  return {
    movie: state.userPreferences.selectedMovie,
    searchValue: state.userPreferences.search.selected
  };
};

export default connect(mapStateToProps)(MovieDetails as ComponentType<any>);
