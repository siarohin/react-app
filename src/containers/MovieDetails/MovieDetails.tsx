import React, { ComponentType, ReactElement, useEffect, useRef } from "react";
import { connect } from "react-redux";

import "./MovieDetails.scss";
import { IconButton, SearchIcon } from "../../shared";
import { MoviesModels, UserPreferencesActions, SharedModels } from "../../core";
import { getFullYear } from "../../utils";
import { IMovieDetails } from "../../models";

const MovieDetails = (props: IMovieDetails & SharedModels.IDispatchAction): ReactElement<IMovieDetails> => {
  const { movie, dispatch } = props;

  const containerRef = useRef((null as unknown) as HTMLDivElement);

  useEffect(() => containerRef.current?.scrollIntoView({ behavior: "smooth" }), [movie]);

  const onClose = (): void => {
    const selectedMovie: MoviesModels.IMovie = {} as MoviesModels.IMovie;
    dispatch(UserPreferencesActions.updateSelectedMovie({ selectedMovie }));
  };

  return (
    <>
      <div ref={containerRef} className="app-header__top-right">
        <IconButton className="app-movie-details__icon-button" onClick={onClose}>
          <SearchIcon />
        </IconButton>
      </div>
      <div className="app-movie-details">
        <div className="app-movie-details__poster" style={{ backgroundImage: "url(" + movie?.posterPath + ")" }}></div>
        <div className="app-movie-detils__content">
          <h3 className="app-movie-details__title">{movie?.title}</h3>
          {movie?.voteAverage ? <div className="app-movie-details__rate">{movie?.voteAverage}</div> : null}
          <div className="app-movie-details__pre-description">
            <span className="app-movie-details__pre-description-item">{getFullYear(movie?.releaseDate)}</span>
            <span className="app-movie-details__pre-description-item">{movie?.runtime} min</span>
          </div>
          <div className="app-movie-details__description">{movie?.overview}</div>
        </div>
      </div>
    </>
  );
};

export default connect()(MovieDetails as ComponentType<any>);
