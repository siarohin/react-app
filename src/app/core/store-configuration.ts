import { applyMiddleware, createStore, compose, Store } from "redux";
import { combineEpics, createEpicMiddleware, Epic } from "redux-observable";

import { reducers, State } from "./store";
import { isServer } from "./constants";

function getComposeEnhancers(): any {
  if (!isServer) {
    return (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
  }
  return compose;
}

export interface AppStore {
  store: Store;
  runEffects: (effects$?: Array<Epic>) => void;
}

/**
 * Store configuration
 */
export function configureStore(initialState: State = {} as State): AppStore {
  const epicMiddleware = createEpicMiddleware();
  const composeEnhancers = getComposeEnhancers();
  const middlewares = [epicMiddleware];

  const store: Store = createStore(reducers, initialState, composeEnhancers(applyMiddleware(...middlewares)));

  return {
    store,
    runEffects: (effects$?: Array<Epic>) =>
      effects$
        ? epicMiddleware.run(combineEpics(...effects$))
        : // eslint-disable-next-line no-console
          console.warn("Effects are not running, no epics array was provided")
  };
}
