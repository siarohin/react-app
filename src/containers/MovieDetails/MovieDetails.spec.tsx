import React from "react";
import { Provider } from "react-redux";
import renderer, { ReactTestRenderer } from "react-test-renderer";
import configureStore from "redux-mock-store";

import { State } from "../../core";
import MovieDetails from "./MovieDetails";

const mockStore = configureStore([]);

describe("Containers.MovieDetails.MovieDetails: ", () => {
  let store: any;
  let component: ReactTestRenderer;

  beforeEach(() => {
    store = mockStore({} as State);

    component = renderer.create(
      <Provider store={store}>
        <MovieDetails movie={undefined} />
      </Provider>
    );
  });

  it("should render component", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});
