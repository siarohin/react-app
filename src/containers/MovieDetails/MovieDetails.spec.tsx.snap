// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers.MovieDetails.MovieDetails:  should render component 1`] = `
Array [
  <div
    className="app-header__top-right"
  >
    <button
      className="MuiButtonBase-root MuiIconButton-root app-movie-details__icon-button"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </span>
    </button>
  </div>,
  <div
    className="app-movie-details"
  >
    <div
      className="app-movie-details__poster"
      style={
        Object {
          "backgroundImage": "url(undefined)",
        }
      }
    />
    <div
      className="app-movie-detils__content"
    >
      <h3
        className="app-movie-details__title"
      />
      <div
        className="app-movie-details__pre-description"
      >
        <span
          className="app-movie-details__pre-description-item"
        >
          
        </span>
        <span
          className="app-movie-details__pre-description-item"
        >
           min
        </span>
      </div>
      <div
        className="app-movie-details__description"
      />
    </div>
  </div>,
]
`;
